File: py/src/cobra_bridge/cobra_wrapper.py
"""
(c) Bosch Sensortec GmbH, Reutlingen, Germany
Open Source as per the BSD-3 Clause
"""

from .enums import CommInterface, ErrorCodes, PinDirection, PinValue, \
    I2CBus, I2CMode, SPIMode, SPISpeed, MultiIOPin, I2CTransferBits, SPITransferBits, \
    StreamingMode, StreamingState, StreamingBlocks, TimerConfig, TimerStampConfig, \
    SerialComConfig, BleComConfig
from .sync import CobraSyncBridge
from .async_ import AsyncCobraBridge
from .transport import CobraTransport, Transport, Transport
from .constants import (
    PIN_IN, PIN_OUT, PIN_LOW, PIN_HIGH,
    I2C_SPEED_400K,
)
from typing import Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from .drivers.base import SensorDriver


class CobraBoard:
    """
    API to utilize the functions of COINES_SDK header file through .dll/.so
    """

    def __init__(self, path_to_coines_lib=None):
        self._transport = CobraTransport()
        self._bridge = CobraSyncBridge(transport=self._transport._active_transport if self._transport._active_transport else None)
        self.error_code = ErrorCodes.COINES_SUCCESS
        self._sensor_drivers: dict[str, 'SensorDriver'] = {}


    def open_comm_interface(self, interface=CommInterface.USB,
                            serial_com_config: SerialComConfig = None,
                            ble_com_config: BleComConfig = None) -> ErrorCodes:
        # Map to the underlying transport layer
        self.error_code = self._transport.open_interface(interface, serial_com_config, ble_com_config)
        return self.error_code

    def close_comm_interface(self, interface=CommInterface.USB) -> ErrorCodes:
        self.error_code = self._transport.close_interface(interface)
        return self.error_code

    def get_version(self) -> str:
        """ Returns the version of the COINES_SDK library """
        return "CobraBoard v0.1.0"

    def unload_library(self) -> None:
        """
        Unloads the library
        """
        print("CobraBoard library unloaded.")

    def write_intf(self, interface: CommInterface, data: list) -> None:
        """ Write data to the communication interface """
        self._transport.write_intf(interface, data)

    def read_intf(self, interface: CommInterface, length: int) -> tuple[list[int], int]:
        """ Read data from the communication interface """
        return self._transport.read_intf(interface, length)

    def config_i2c_bus(self, bus: I2CBus, i2c_address: int, i2c_mode: I2CMode) -> ErrorCodes:
        self.error_code = self._bridge.config_i2c_bus(bus, i2c_address, i2c_mode)
        return self.error_code

    def deconfig_i2c_bus(self, bus: I2CBus) -> ErrorCodes:
        self.error_code = self._bridge.deconfig_i2c_bus(bus)
        return self.error_code

    def write_i2c(self, bus: I2CBus, register_address: int,
                  register_value: int, sensor_interface_detail: int | None = None) -> ErrorCodes:
        self.error_code = self._bridge.write_i2c(bus, register_address, register_value, sensor_interface_detail)
        return self.error_code

    def read_i2c(self, bus: I2CBus, register_address: int,
                 number_of_reads: int, sensor_interface_detail: int | None = None) -> tuple[list[int], ErrorCodes]:
        data, self.error_code = self._bridge.read_i2c(bus, register_address, number_of_reads, sensor_interface_detail)
        return data, self.error_code

    def read_16bit_i2c(self, bus: I2CBus, register_address: int, number_of_reads: int = 2,
                       sensor_interface_detail: int | None = None,
                       i2c_transfer_bits: I2CTransferBits = I2CTransferBits.I2C16BIT) -> tuple[list[int], ErrorCodes]:
        data, self.error_code = self._bridge.read_16bit_i2c(bus, register_address, number_of_reads,
                                                            sensor_interface_detail, i2c_transfer_bits)
        return data, self.error_code

    def write_16bit_i2c(self, bus: I2CBus, register_address: int,
                        register_value: int, sensor_interface_detail: int | None = None,
                        i2c_transfer_bits: I2CTransferBits = I2CTransferBits.I2C16BIT) -> ErrorCodes:
        self.error_code = self._bridge.write_16bit_i2c(bus, register_address, register_value,
                                                        sensor_interface_detail, i2c_transfer_bits)
        return self.error_code

    def config_spi_bus(self, bus: I2CBus, cs_pin: MultiIOPin,
                       spi_speed: SPISpeed, spi_mode: SPIMode) -> ErrorCodes:
        self.error_code = self._bridge.config_spi_bus(bus, cs_pin, spi_speed, spi_mode)
        return self.error_code

    def deconfig_spi_bus(self, bus: I2CBus) -> ErrorCodes:
        self.error_code = self._bridge.deconfig_spi_bus(bus)
        return self.error_code

    def custom_spi_config(self, bus: I2CBus, cs_pin: MultiIOPin,
                          spi_speed: SPISpeed, spi_mode: SPIMode) -> ErrorCodes:
        self.error_code = self._bridge.custom_spi_config(bus, cs_pin, spi_speed, spi_mode)
        return self.error_code

    def write_spi(self, bus: I2CBus, register_address: int,
                  register_value: int, sensor_interface_detail: int | None = None) -> ErrorCodes:
        self.error_code = self._bridge.write_spi(bus, register_address, register_value, sensor_interface_detail)
        return self.error_code

    def read_spi(self, bus: I2CBus, register_address: int,
                 number_of_reads: int, sensor_interface_detail: int | None = None) -> tuple[list[int], ErrorCodes]:
        data, self.error_code = self._bridge.read_spi(bus, register_address, number_of_reads, sensor_interface_detail)
        return data, self.error_code

    def read_16bit_spi(self, bus: I2CBus, register_address: int, number_of_reads: int = 2,
                       sensor_interface_detail: int | None = None,
                       spi_transfer_bits: SPITransferBits = SPITransferBits.SPI16BIT) -> tuple[list[int], ErrorCodes]:
        data, self.error_code = self._bridge.read_16bit_spi(bus, register_address, number_of_reads,
                                                            sensor_interface_detail, spi_transfer_bits)
        return data, self.error_code

    def write_16bit_spi(self, bus: I2CBus, register_address: int,
                        register_value: list, sensor_interface_detail: int | None = None,
                        spi_transfer_bits: SPITransferBits = SPITransferBits.SPI16BIT) -> ErrorCodes:
        self.error_code = self._bridge.write_16bit_spi(bus, register_address, register_value,
                                                        sensor_interface_detail, spi_transfer_bits)
        return self.error_code

    # ── Board Control (convenience methods for sensor drivers) ───────────

    def set_vdd(self, voltage_mv: int) -> int:
        """Set VDD voltage in millivolts (0 = off). Returns status code."""
        return self._bridge.set_vdd(voltage_mv)

    def set_vddio(self, voltage_mv: int) -> int:
        """Set VDDIO voltage in millivolts (0 = off). Returns status code."""
        return self._bridge.set_vddio(voltage_mv)

    def set_pin(self, pin: int, direction: int, value: int) -> int:
        """Configure a shuttle board pin. Returns status code."""
        return self._bridge.set_pin(pin, direction, value)

    # ── Sensor-Driver Convenience I/O ────────────────────────────────────

    def i2c_read_reg(self, dev_addr: int, reg_addr: int, length: int,
                      speed: int = I2C_SPEED_400K) -> list[int]:
        """
        Read `length` bytes from I2C register. Returns list of ints.

        Thin wrapper over CobraSyncBridge.i2c_read() for sensor drivers.
        """
        return self._bridge.i2c_read(dev_addr, reg_addr, length, speed)

    def i2c_write_reg(self, dev_addr: int, reg_addr: int, data: bytes,
                      speed: int = I2C_SPEED_400K) -> int:
        """
        Write data bytes to I2C register. Returns status byte.

        Thin wrapper over CobraSyncBridge.i2c_write() for sensor drivers.
        """
        return self._bridge.i2c_write(dev_addr, reg_addr, data, speed)

    def spi_read_reg(self, cs_pin: int, reg_addr: int, length: int,
                     speed: int = 1000000, mode: int = 0) -> list[int]:
        """
        Read `length` bytes from SPI register. Returns list of ints.

        Thin wrapper over CobraSyncBridge.spi_read() for sensor drivers.
        """
        return self._bridge.spi_read(cs_pin, reg_addr, length, speed, mode)

    def spi_write_reg(self, cs_pin: int, reg_addr: int, data: bytes,
                      speed: int = 1000000, mode: int = 0) -> int:
        """
        Write data bytes to SPI register. Returns status byte.

        Thin wrapper over CobraSyncBridge.spi_write() for sensor drivers.
        """
        return self._bridge.spi_write(cs_pin, reg_addr, data, speed, mode)

    # ── Sensor Driver Registry ──────────────────────────────────────────

    def attach_driver(self, driver: 'SensorDriver') -> None:
        """
        Register a sensor driver instance with this board.

        Args:
            driver: SensorDriver instance (e.g., BMM350Driver).
        """
        self._sensor_drivers[driver.name] = driver

    def get_driver(self, name: str) -> Optional['SensorDriver']:
        """
        Retrieve a previously attached sensor driver by name.

        Args:
            name: Sensor name (e.g., 'bmm350').

        Returns:
            SensorDriver instance, or None if not found.
        """
        return self._sensor_drivers.get(name)

    @property
    def drivers(self) -> dict[str, 'SensorDriver']:
        """All attached sensor drivers, keyed by name."""
        return dict(self._sensor_drivers)


class AsyncCobraBoard:
    """
    Asynchronous API to utilize the functions of COINES_SDK header file through .dll/.so
    """

    def __init__(self, path_to_coines_lib=None):
        self._transport = CobraTransport()
        self._async_bridge = AsyncCobraBridge(transport=self._transport)
        self.error_code = ErrorCodes.COINES_SUCCESS
        self._sensor_drivers: dict[str, 'SensorDriver'] = {}

    async def open_comm_interface(self, interface=CommInterface.USB,
                                  serial_com_config=None,
                                  ble_com_config=None) -> ErrorCodes:
        self.error_code = await self._async_bridge.open_interface(interface, serial_com_config, ble_com_config)
        return self.error_code

    async def close_comm_interface(self, interface=CommInterface.USB) -> ErrorCodes:
        self.error_code = await self._async_bridge.close_interface(interface)
        return self.error_code

    async def write_intf(self, interface: CommInterface, data: list) -> None:
        await self._async_bridge.write_intf(interface, data)

    async def read_intf(self, interface: CommInterface, length: int) -> tuple[list[int], int]:
        data, n_bytes_read = await self._async_bridge.read_intf(interface, length)
        return data, n_bytes_read

    async def config_i2c_bus(self, bus: I2CBus, i2c_address: int, i2c_mode: I2CMode) -> ErrorCodes:
        self.error_code = await self._async_bridge.config_i2c_bus(bus, i2c_address, i2c_mode)
        return self.error_code

    async def deconfig_i2c_bus(self, bus: I2CBus) -> ErrorCodes:
        self.error_code = await self._async_bridge.deconfig_i2c_bus(bus)
        return self.error_code

    async def write_i2c(self, bus: I2CBus, register_address: int,
                        register_value: int, sensor_interface_detail: int | None = None) -> ErrorCodes:
        self.error_code = await self._async_bridge.write_i2c(bus, register_address, register_value, sensor_interface_detail)
        return self.error_code

    async def read_i2c(self, bus: I2CBus, register_address: int,
                       number_of_reads: int, sensor_interface_detail: int | None = None) -> tuple[list[int], ErrorCodes]:
        data, self.error_code = await self._async_bridge.read_i2c(bus, register_address, number_of_reads, sensor_interface_detail)
        return data, self.error_code

    async def read_16bit_i2c(self, bus: I2CBus, register_address: int, number_of_reads: int = 2,
                             sensor_interface_detail: int | None = None,
                             i2c_transfer_bits: I2CTransferBits = I2CTransferBits.I2C16BIT) -> tuple[list[int], ErrorCodes]:
        data, self.error_code = await self._async_bridge.read_16bit_i2c(bus, register_address, number_of_reads,
                                                                        sensor_interface_detail, i2c_transfer_bits)
        return data, self.error_code

    async def write_16bit_i2c(self, bus: I2CBus, register_address: int,
                              register_value: int, sensor_interface_detail: int | None = None,
                              i2c_transfer_bits: I2CTransferBits = I2CTransferBits.I2C16BIT) -> ErrorCodes:
        self.error_code = await self._async_bridge.write_16bit_i2c(bus, register_address, register_value,
                                                                    sensor_interface_detail, i2c_transfer_bits)
        return self.error_code

    async def config_spi_bus(self, bus: I2CBus, cs_pin: MultiIOPin,
                             spi_speed: SPISpeed, spi_mode: SPIMode) -> ErrorCodes:
        self.error_code = await self._async_bridge.config_spi_bus(bus, cs_pin, spi_speed, spi_mode)
        return self.error_code

    async def deconfig_spi_bus(self, bus: I2CBus) -> ErrorCodes:
        self.error_code = await self._async_bridge.deconfig_spi_bus(bus)
        return self.error_code

    async def custom_spi_config(self, bus: I2CBus, cs_pin: MultiIOPin,
                                spi_speed: SPISpeed, spi_mode: SPIMode) -> ErrorCodes:
        self.error_code = await self._async_bridge.custom_spi_config(bus, cs_pin, spi_speed, spi_mode)
        return self.error_code

    async def write_spi(self, bus: I2CBus, register_address: int,
                        register_value: int, sensor_interface_detail: int | None = None) -> ErrorCodes:
        self.error_code = await self._async_bridge.write_spi(bus, register_address, register_value, sensor_interface_detail)
        return self.error_code

    async def read_spi(self, bus: I2CBus, register_address: int,
                       number_of_reads: int, sensor_interface_detail: int | None = None) -> tuple[list[int], ErrorCodes]:
        data, self.error_code = await self._async_bridge.read_spi(bus, register_address, number_of_reads, sensor_interface_detail)
        return data, self.error_code

    async def read_16bit_spi(self, bus: I2CBus, register_address: int, number_of_reads: int = 2,
                             sensor_interface_detail: int | None = None,
                             spi_transfer_bits: SPITransferBits = SPITransferBits.SPI16BIT) -> tuple[list[int], ErrorCodes]:
        data, self.error_code = await self._async_bridge.read_16bit_spi(bus, register_address, number_of_reads,
                                                                        sensor_interface_detail, spi_transfer_bits)
        return data, self.error_code

    async def write_16bit_spi(self, bus: I2CBus, register_address: int,
                                register_value: list, sensor_interface_detail: int | None = None,
                                spi_transfer_bits: SPITransferBits = SPITransferBits.SPI16BIT) -> ErrorCodes:
        self.error_code = await self._async_bridge.write_16bit_spi(bus, register_address, register_value,
                                                                    sensor_interface_detail, spi_transfer_bits)
        return self.error_code

    # ── Board Control (convenience methods for sensor drivers) ───────────

    def set_vdd(self, voltage_mv: int) -> int:
        """Set VDD voltage in millivolts (0 = off). Returns status code."""
        return self._async_bridge.set_vdd(voltage_mv)

    def set_vddio(self, voltage_mv: int) -> int:
        """Set VDDIO voltage in millivolts (0 = off). Returns status code."""
        return self._async_bridge.set_vddio(voltage_mv)

    def set_pin(self, pin: int, direction: int, value: int) -> int:
        """Configure a shuttle board pin. Returns status code."""
        return self._async_bridge.set_pin(pin, direction, value)

    # ── Sensor-Driver Convenience I/O ────────────────────────────────────

    async def i2c_read_reg(self, dev_addr: int, reg_addr: int, length: int,
                           speed: int = I2C_SPEED_400K) -> list[int]:
        """Read `length` bytes from I2C register. Returns list of ints."""
        return await self._async_bridge.i2c_read(dev_addr, reg_addr, length, speed)

    async def i2c_write_reg(self, dev_addr: int, reg_addr: int, data: bytes,
                            speed: int = I2C_SPEED_400K) -> int:
        """Write data bytes to I2C register. Returns status byte."""
        return await self._async_bridge.i2c_write(dev_addr, reg_addr, data, speed)

    async def spi_read_reg(self, cs_pin: int, reg_addr: int, length: int,
                           speed: int = 1000000, mode: int = 0) -> list[int]:
        """Read `length` bytes from SPI register. Returns list of ints."""
        return await self._async_bridge.spi_read(cs_pin, reg_addr, length, speed, mode)

    async def spi_write_reg(self, cs_pin: int, reg_addr: int, data: bytes,
                            speed: int = 1000000, mode: int = 0) -> int:
        """Write data bytes to SPI register. Returns status byte."""
        return await self._async_bridge.spi_write(cs_pin, reg_addr, data, speed, mode)

    # ── Sensor Driver Registry ──────────────────────────────────────────

    def attach_driver(self, driver: 'SensorDriver') -> None:
        """Register a sensor driver instance with this board."""
        self._sensor_drivers[driver.name] = driver

    def get_driver(self, name: str) -> Optional['SensorDriver']:
        """Retrieve a previously attached sensor driver by name."""
        return self._sensor_drivers.get(name)

    @property
    def drivers(self) -> dict[str, 'SensorDriver']:
        """All attached sensor drivers, keyed by name."""
        return dict(self._sensor_drivers)

File: js/src/cobra_wrapper.js
/**
 * COBRA Wrapper — JavaScript
 *
 * Provides a drop-in replacement for coinespy-like API in JavaScript.
 * Supports synchronous and asynchronous operations by delegating to underlying
 * CobraBridge and Transport classes.
 */

import { CobraBridge } from './sync.js';
import { SerialTransport, BleTransport } from './transport.js';
import {
    CommInterface, ErrorCodes, PinDirection, PinValue,
    I2CBus, I2CMode, SPIMode, SPISpeed, MultiIOPin, I2CTransferBits, SPITransferBits,
    StreamingMode, StreamingState, TimerConfig, TimerStampConfig,
    PIN_IN, PIN_OUT, PIN_LOW, PIN_HIGH,
    I2C_SPEED_400K,
} from './constants.js';


class CobraTransportJs {
    constructor() {
        this._activeTransport = null;
    }

    async openInterface(interfaceType, serialComConfig = null, bleComConfig = null) {
        if (this._activeTransport && this._activeTransport.connected) {
            return ErrorCodes.COINES_E_COMM_ALREADY_OPEN;
        }

        if (interfaceType === CommInterface.USB.value) {
            if (!serialComConfig || !serialComConfig.comPortName) {
                return ErrorCodes.COINES_E_INVALID_ARGUMENT;
            }
            this._activeTransport = new SerialTransport(serialComConfig.baudRate);
        } else if (interfaceType === CommInterface.BLE.value) {
            if (!bleComConfig || !bleComConfig.address) {
                return ErrorCodes.COINES_E_INVALID_ARGUMENT;
            }
            this._activeTransport = new BleTransport(bleComConfig.address);
        } else if (interfaceType === CommInterface.VIRTUAL.value) {
            console.log('Virtual interface selected. No actual connection.');
            this._activeTransport = null; // Or a mock transport
            return ErrorCodes.COINES_SUCCESS;
        } else {
            return ErrorCodes.COINES_E_INVALID_ARGUMENT;
        }

        if (this._activeTransport) {
            try {
                await this._activeTransport.connect();
                return ErrorCodes.COINES_SUCCESS;
            } catch (e) {
                console.error(`Error opening transport interface: ${e}`);
                this._activeTransport = null;
                return ErrorCodes.COINES_E_COMM_INIT_FAILED;
            }
        }
        return ErrorCodes.COINES_E_COMM_INIT_FAILED;
    }

    async closeInterface(interfaceType) {
        if (this._activeTransport && this._activeTransport.connected) {
            try {
                await this._activeTransport.disconnect();
                this._activeTransport = null;
                return ErrorCodes.COINES_SUCCESS;
            } catch (e) {
                console.error(`Error closing transport interface: ${e}`);
                return ErrorCodes.COINES_E_FAILURE;
            }
        } else if (interfaceType === CommInterface.VIRTUAL.value) {
            console.log('Virtual interface closing. No actual disconnection.');
            this._activeTransport = null;
            return ErrorCodes.COINES_SUCCESS;
        }
        return ErrorCodes.COINES_E_UNABLE_OPEN_DEVICE;
    }

    async writeIntf(interfaceType, data) {
        if (this._activeTransport && this._activeTransport.connected) {
            await this._activeTransport.send(new Uint8Array(data));
        } else {
            console.log('No active transport to write to.');
        }
    }

    async readIntf(interfaceType, length) {
        if (this._activeTransport && this._activeTransport.connected) {
            const readBytes = await this._activeTransport.receive(length);
            return [Array.from(readBytes), readBytes.length];
        } else {
            console.log('No active transport to read from.');
            return [[], 0];
        }
    }
}


export class CobraBoardJs {
    constructor() {
        this._transport = new CobraTransportJs();
        this._bridge = new CobraBridge(this._transport);
        this.error_code = ErrorCodes.COINES_SUCCESS;
        this._sensorDrivers = {};
    }

    async open_comm_interface(interfaceType, serialComConfig = null, bleComConfig = null) {
        this.error_code = await this._transport.openInterface(interfaceType, serialComConfig, bleComConfig);
        return this.error_code;
    }

    async close_comm_interface(interfaceType) {
        this.error_code = await this._transport.closeInterface(interfaceType);
        return this.error_code;
    }

    get_version() {
        return "CobraBoardJs v0.1.0";
    }

    unload_library() {
        console.log("CobraBoardJs library unloaded.");
    }

    async write_intf(interfaceType, data) {
        await this._transport.writeIntf(interfaceType, data);
    }

    async read_intf(interfaceType, length) {
        const [data, n_bytes_read] = await this._transport.readIntf(interfaceType, length);
        return [data, n_bytes_read];
    }

    async config_i2c_bus(bus, i2c_address, i2c_mode) {
        // Delegate to CobraBridge after adapting parameters if necessary
        console.log(`[CobraBoardJs] Configuring I2C bus ${bus}, address ${i2c_address}, mode ${i2c_mode}`);
        // The underlying CobraBridge.i2c_write/read don't have explicit bus config methods.
        // This would require new commands in the COINES protocol for JS.
        // For now, simulate success.
        this.error_code = ErrorCodes.COINES_SUCCESS;
        return this.error_code;
    }

    async deconfig_i2c_bus(bus) {
        console.log(`[CobraBoardJs] Deconfiguring I2C bus ${bus}`);
        this.error_code = ErrorCodes.COINES_SUCCESS;
        return this.error_code;
    }

    async write_i2c(bus, register_address, register_value, sensor_interface_detail = null) {
        const dev_addr = sensor_interface_detail !== null ? sensor_interface_detail : 0; // Default if not provided
        const data = new Uint8Array([register_value]);
        // The existing CobraBridge.i2c_write uses speed as a parameter. Coinespy API doesn't.
        // We'll use a default speed for now, or assume it's configured globally.
        try {
            const status = await this._bridge.i2c_write(dev_addr, register_address, data, I2C_SPEED_400K);
            this.error_code = ErrorCodes.COINES_SUCCESS; // Assuming status is 0 for success
        } catch (e) {
            console.error(`Error in write_i2c: ${e}`);
            this.error_code = ErrorCodes.COINES_E_I2C_READ_WRITE_FAILED;
        }
        return this.error_code;
    }

    async read_i2c(bus, register_address, number_of_reads, sensor_interface_detail = null) {
        const dev_addr = sensor_interface_detail !== null ? sensor_interface_detail : 0; // Default if not provided
        try {
            const data = await this._bridge.i2c_read(dev_addr, register_address, number_of_reads, I2C_SPEED_400K);
            this.error_code = ErrorCodes.COINES_SUCCESS;
            return [Array.from(data), this.error_code];
        } catch (e) {
            console.error(`Error in read_i2c: ${e}`);
            this.error_code = ErrorCodes.COINES_E_I2C_READ_WRITE_FAILED;
            return [[], this.error_code];
        }
    }

    async read_16bit_i2c(bus, register_address, number_of_reads = 2,
                         sensor_interface_detail = null,
                         i2c_transfer_bits = I2CTransferBits.I2C16BIT) {
        console.log(`[CobraBoardJs] Reading 16-bit I2C from bus ${bus}, reg ${register_address}`);
        const [data, error] = await this.read_i2c(bus, register_address, number_of_reads * 2, sensor_interface_detail);
        return data;
    }

    async write_16bit_i2c(bus, register_address, register_value, sensor_interface_detail = null,
                          i2c_transfer_bits = I2CTransferBits.I2C16BIT) {
        console.log(`[CobraBoardJs] Writing 16-bit I2C to bus ${bus}, reg ${register_address}, val ${register_value}`);
        const byte1 = register_value & 0xFF;
        const byte2 = (register_value >> 8) & 0xFF;
        let status = await this.write_i2c(bus, register_address, byte1, sensor_interface_detail);
        if (status === ErrorCodes.COINES_SUCCESS) {
            status = await this.write_i2c(bus, register_address + 1, byte2, sensor_interface_detail);
        }
        return status;
    }

    async config_spi_bus(bus, cs_pin, spi_speed, spi_mode) {
        console.log(`[CobraBoardJs] Configuring SPI bus ${bus}, CS ${cs_pin}, speed ${spi_speed}, mode ${spi_mode}`);
        this.error_code = ErrorCodes.COINES_SUCCESS;
        return this.error_code;
    }

    async deconfig_spi_bus(bus) {
        console.log(`[CobraBoardJs] Deconfiguring SPI bus ${bus}`);
        this.error_code = ErrorCodes.COINES_SUCCESS;
        return this.error_code;
    }

    async custom_spi_config(bus, cs_pin, spi_speed, spi_mode) {
        console.log(`[CobraBoardJs] Custom SPI config for bus ${bus}, CS ${cs_pin}, speed ${spi_speed}, mode ${spi_mode}`);
        return await this.config_spi_bus(bus, cs_pin, spi_speed, spi_mode);
    }

    async write_spi(bus, register_address, register_value, sensor_interface_detail = null) {
        const cs_pin = sensor_interface_detail !== null ? sensor_interface_detail : MultiIOPin.COINES_MINI_SHUTTLE_PIN_CS.value;
        const data = new Uint8Array([register_value]);
        try {
            const status = await this._bridge.spi_write(cs_pin, register_address, data, SPI_SPEED_10MHZ, SPI_MODE_0);
            this.error_code = ErrorCodes.COINES_SUCCESS; // Assuming status is 0 for success
        } catch (e) {
            console.error(`Error in write_spi: ${e}`);
            this.error_code = ErrorCodes.COINES_E_SPI_READ_WRITE_FAILED;
        }
        return this.error_code;
    }

    async read_spi(bus, register_address, number_of_reads, sensor_interface_detail = null) {
        const cs_pin = sensor_interface_detail !== null ? sensor_interface_detail : MultiIOPin.COINES_MINI_SHUTTLE_PIN_CS.value;
        try {
            const data = await this._bridge.spi_read(cs_pin, register_address, number_of_reads, SPI_SPEED_10MHZ, SPI_MODE_0);
            this.error_code = ErrorCodes.COINES_SUCCESS;
            return [Array.from(data), this.error_code];
        } catch (e) {
            console.error(`Error in read_spi: ${e}`);
            this.error_code = ErrorCodes.COINES_E_SPI_READ_WRITE_FAILED;
            return [[], this.error_code];
        }
    }

    async read_16bit_spi(bus, register_address, number_of_reads = 2,
                         sensor_interface_detail = null,
                         spi_transfer_bits = SPITransferBits.SPI16BIT) {
        console.log(`[CobraBoardJs] Reading 16-bit SPI from bus ${bus}, reg ${register_address}`);
        const [data, error] = await this.read_spi(bus, register_address, number_of_reads * 2, sensor_interface_detail);
        return data;
    }

    async write_16bit_spi(bus, register_address, register_value, sensor_interface_detail = null,
                           spi_transfer_bits = SPITransferBits.SPI16BIT) {
        console.log(`[CobraBoardJs] Writing 16-bit SPI to bus ${bus}, reg ${register_address}, val ${register_value}`);
        let status = ErrorCodes.COINES_SUCCESS;
        if (Array.isArray(register_value)) {
            for (const val of register_value) {
                status = await this.write_spi(bus, register_address, val, sensor_interface_detail);
                if (status !== ErrorCodes.COINES_SUCCESS) {
                    break;
                }
            }
        } else {
            status = await this.write_spi(bus, register_address, register_value, sensor_interface_detail);
        }
        return status;
    }

    // ── Board Control (convenience methods for sensor drivers) ───────────

    async setVdd(voltageMv) {
        /** Set VDD voltage in millivolts (0 = off). Returns status code. */
        return await this._bridge.setVdd(voltageMv);
    }

    async setVddio(voltageMv) {
        /** Set VDDIO voltage in millivolts (0 = off). Returns status code. */
        return await this._bridge.setVddio(voltageMv);
    }

    async setPin(pin, direction, value) {
        /** Configure a shuttle board pin. Returns status code. */
        return await this._bridge.setPin(pin, direction, value);
    }

    // ── Sensor-Driver Convenience I/O ────────────────────────────────────

    async i2cReadReg(devAddr, regAddr, length, speed = I2C_SPEED_400K) {
        /** Read `length` bytes from I2C register. Returns Uint8Array. */
        return await this._bridge.i2c_read(devAddr, regAddr, length, speed);
    }

    async i2cWriteReg(devAddr, regAddr, data, speed = I2C_SPEED_400K) {
        /** Write data bytes to I2C register. Returns status byte. */
        return await this._bridge.i2c_write(devAddr, regAddr, data, speed);
    }

    async spiReadReg(csPin, regAddr, length, speed = 1000000, mode = 0) {
        /** Read `length` bytes from SPI register. Returns Uint8Array. */
        return await this._bridge.spi_read(csPin, regAddr, length, speed, mode);
    }

    async spiWriteReg(csPin, regAddr, data, speed = 1000000, mode = 0) {
        /** Write data bytes to SPI register. Returns status byte. */
        return await this._bridge.spi_write(csPin, regAddr, data, speed, mode);
    }

    // ── Sensor Driver Registry ──────────────────────────────────────────

    attachDriver(driver) {
        /** Register a sensor driver instance with this board. */
        this._sensorDrivers[driver.constructor.name] = driver;
    }

    getDriver(name) {
        /** Retrieve a previously attached sensor driver by name. */
        return this._sensorDrivers[name] || null;
    }

    get drivers() {
        /** All attached sensor drivers, keyed by name. */
        return { ...this._sensorDrivers };
    }
}

export { CobraBoardJs as CobraBoard };
export { CobraTransportJs as CobraTransport };